src/models/session/session-state.ts

```diff
@@ -1,187 +1,110 @@
-    breadcrumbs.forEach((crumb: NormalizedCrumb) => {
-      const notSeenThisSection =
-        !seenCrumb || (seenCrumb && crumb.sectionId != seenCrumb.sectionId);
-      if (notSeenThisSection) {
-        // increment section index
-        sectionIndex++;
-        // mark the section as ReadyToStart
-        if (sections[sectionIndex]) {
-          sections[sectionIndex].status = SectionStatuses.ReadyToStart;
-        }
-        // mark the previous section as Complete
-        sections[sectionIndex - 1].status = SectionStatuses.Complete;
-      }
-      // track this crumb as seen
-      seenCrumb = crumb;
-    });
-
-    // compute ReadyToContinue and CannotContinueYet statuses
-    if (cachedBreadcrumbs) {
-      const upcomingIds = this.remainingNodes(breadcrumbs).map(
-        (node: NormalizedNode) => node.id
-      );
-      cachedBreadcrumbs
-        .filter((crumb) => upcomingIds.includes(crumb.id))
-        .reverse() // work backwards to prioritize ReadyToContinue
-        .forEach((crumb) => {
-          const sectionIndex = sections.findIndex(
-            (section) => section.id == crumb.sectionId
-          );
-          sections[sectionIndex].status = SectionStatuses.CannotContinueYet;
-          if (upcomingIds[0] === crumb.id) {
-            sections[sectionIndex].status = SectionStatuses.ReadyToContinue;
-          }
-        });
+      return this.structuredFlow;
     }
+    const lastCrumb = breadcrumbs.at(-1)!;
+    const indexedNode = this.findIndexedNode(lastCrumb.id);
+    if (!indexedNode)
+      throw new Error(`Breadcrumb node "${lastCrumb.id}" could not be found`);
+    const branchIndex = this.structuredFlow.findIndex(
+      (branch: StructuredNode) => branch.id === indexedNode!.rootNodeId
+    );
+    return this.structuredFlow.slice(branchIndex + 1);
+  }
 
-    // compute the Updated status
-    if (updatedNodeIds) {
-      breadcrumbs.forEach((crumb) => {
-        [crumb.id, ...(crumb.answers || [])]
-          .filter((answerId) => updatedNodeIds.includes(answerId))
-          .forEach((id) => {
-            if (updatedNodeIds.includes(id)) {
-              const sId = sections.findIndex((s) => s.id == crumb.sectionId);
-              sections[sId].status = SectionStatuses.Updated;
-            }
-          });
-      });
-    }
+  private findIndexedNode(id: NodeId): IndexedNode | undefined {
+    return this.orderedFlow.find((node) => node.id === id);
+  }
 
-    return sections;
+  private findLastIndexedNode(id: NodeId): IndexedNode | undefined {
+    return this.orderedFlow.reverse().find((node) => node.id === id);
   }
 
-  private currentIndex(breadcrumbs: OrderedBreadcrumbs) {
-    const lastCrumb: NormalizedCrumb = breadcrumbs.at(-1)!;
-    const currentIndex = this.normalizedFlow.findIndex((node) => {
-      if (lastCrumb.answers?.length) {
-        const lastAnswerId = lastCrumb.answers.at(-1);
-        return node.id === lastAnswerId;
+  private findStructuredNode(
```

Comment:
nit: This function is clearly pretty important and called for multiple operations. I think the naming here is clear and meaningful but I'd personally be inclined to add a few comments just to be super explicit about what each step is doing with long-term maintenance in mind.

Reply:
I appreciate this a bit contradictory sorry and we shouldn't need comments when it's clear and meaningful, but I guess I'm thinking of future-proofing here as best we can.